src/Mutator/MutatorFactory.php

```diff
@@ -166,16 +165,19 @@ private static function registerFromClass(
     /**
      * @param array<string, array<string, string>> $mutatorNames
      *
-     * @return array<string, Mutator>
+     * @return array<string, IgnoreMutator>
      */
     private static function createFromNames(array $mutatorNames): array
     {
         $mutators = [];
 
         foreach ($mutatorNames as $mutatorClass => $settings) {
+            $mutatorConfig = new MutatorConfig($settings);
+
             /** @var Mutator $mutator */
-            $mutator = new $mutatorClass(new MutatorConfig($settings));
-            $mutators[$mutator::getName()] = $mutator;
+            $mutator = new $mutatorClass($mutatorConfig);
```

Comment:
I admit it's still a point of confusion to me to be honest. To clear a few things up:

- Right now `schema.json` and `ProfileList` do not allow custom mutators
- The config API is a bit awkward: you are passing a raw object hence requiring to normalize & validate the config (although for now we managed to validate most of it if not all in `schema.json`) in the mutator which IMO is quite late in the process to find out there is a configuration error

Setting those potential custom mutator aside, I was thinking of being able to normalize this config upstream at the `ConfigurationFactory` level (not necessarily in that class but in that process, i.e. the `Configuration` object would have its mutators with a normalized config already). The few mutators (we have 2 or 3 of them) which require a special config to be passed in the constructor still can be handled in `MutatorFactory`.

Now what about custom mutator? I think this should be discussed more, but maybe an idea is to switch from a JSON config file to a PHP one? This way you could have something along the lines of:

```php
<?php

require __DIR__.'/path/to/custom-mutator/autoloader.php;

use Infection\Mutator\MutatorDefinition;
use Infection\Mutator\Mutator;
use Acme\MyMutatorConfig;
use Acme\MyMutator;

return [
     'mutators' => [
         new MutatorDefinition(
             'MyMutator', 
             MyMutator::class, 
             static function (array $settings): Mutator {
                 return new MyMutator(new MyMutatorConfig($settings));
             }
         )
     ],
];
```

We then have all the information needed and that factory provided by the user can be injected & used in the `MutatorFactory`.

It's just an idea, I think we need to think this through. But since this is involving a lot of non-trivial elements (how much of our API are we sharing? how? what about the autoloading? What about when we scope the PHAR? what about name conflicts? If we disable "no more elements" in the schema.json how to make sure we don't miss any new mutator we add?) I didn't want to think to much of it yet.